sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/ConstraintPropagationSuite.scala

```diff
@@ -217,4 +218,12 @@ class ConstraintPropagationSuite extends SparkFunSuite {
         IsNotNull(resolveColumn(tr, "a")),
         IsNotNull(resolveColumn(tr, "b")))))
   }
+
+  test("infer constraints on cast") {
+    val tr = LocalRelation('a.int, 'b.long)
+    verifyConstraints(tr.where('a.attr === 'b.attr).analyze.constraints,
```

Comment:
I think adding `IsNotNull(Cast(Cast(resolveColumn(tr, "a"), LongType), LongType))` is not a valid test case for this. Because as you did this, you explicitly add this into constraints and we don't remove any expression from constraints now. So it will show up in final constraints.